Return-Path: <linux-arm-kernel-bounces+thomas.petazzoni=free-electrons.com@lists.infradead.org>
Delivered-To: thomas.petazzoni@free-electrons.com
Received: by mail.free-electrons.com (Postfix, from userid 106)
	id 6F005142; Wed,  4 Apr 2012 20:37:27 +0200 (CEST)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	mail.free-electrons.com
X-Spam-Level: 
X-Spam-Status: No, score=-5.1 required=1.0 tests=AWL,BAYES_00,
	RCVD_IN_DNSWL_MED,T_RP_MATCHES_RCVD shortcircuit=no autolearn=unavailable
	version=3.3.1
Received: from merlin.infradead.org (merlin.infradead.org [205.233.59.134])
	by mail.free-electrons.com (Postfix) with ESMTPS id A1EDC18A
	for <thomas.petazzoni@free-electrons.com>; Wed,  4 Apr 2012 20:37:13 +0200 (CEST)
Received: from localhost ([::1] helo=merlin.infradead.org)
	by merlin.infradead.org with esmtp (Exim 4.76 #1 (Red Hat Linux))
	id 1SFV5B-0000UI-U3; Wed, 04 Apr 2012 18:37:25 +0000
Received: from mail-we0-f177.google.com ([74.125.82.177])
 by merlin.infradead.org with esmtps (Exim 4.76 #1 (Red Hat Linux))
 id 1SFV52-0000Si-IK
 for linux-arm-kernel@lists.infradead.org; Wed, 04 Apr 2012 18:37:18 +0000
Received: by werp11 with SMTP id p11so441106wer.36
 for <linux-arm-kernel@lists.infradead.org>;
 Wed, 04 Apr 2012 11:37:14 -0700 (PDT)
X-Google-DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
 d=google.com; s=20120113;
 h=from:to:cc:subject:date:message-id:x-mailer:in-reply-to:references
 :x-gm-message-state;
 bh=kNl3ujVgjSkb4W2vsUFhTDThyeE3x+aSl4tmS/+FMUQ=;
 b=aDi5B5BJQwMYcSOFUjaIJete6c7XFKPiL2xgfcNT3fNYk/OamIhOtGiF3TzRJjq4C+
 5SvsW/OKnedwDA69PEi7NVAnV6WgNYMZ7eqscvjo87eBNxqTsfHIDb7+RYk6AQQDQ3VR
 9EBlD6U80oNmtUCJih+zaBt4cpp15zzBqtA8p71vaqPORL5LLPOwl+ReDKjZqo7WDT6n
 YW5snNfVw9INMbYAHC19kmfjXGMiUEwAfm9vjB9+GdZ4sEOiSggtxhDaU2a9Gv0GSY7X
 zKRoVr+naw6RMO9aD7qlB8AfFOoVa1wlEGKropJ5zXqtmd3IZ1FmXu3hEavwv/aSycL9
 zLww==
Received: by 10.216.133.234 with SMTP id q84mr2066602wei.102.1333564633636;
 Wed, 04 Apr 2012 11:37:13 -0700 (PDT)
Received: from localhost.localdomain ([188.77.211.255])
 by mx.google.com with ESMTPS id 17sm7060020wis.0.2012.04.04.11.37.11
 (version=TLSv1/SSLv3 cipher=OTHER);
 Wed, 04 Apr 2012 11:37:12 -0700 (PDT)
From: Javier Martinez Canillas <javier@dowhile0.org>
To: Tony Lindgren <tony@atomide.com>
Subject: [RESEND PATCH 2/2] OMAP3: igep0020: Add support for Micron NAND Flash
 storage memory
Date: Wed,  4 Apr 2012 20:36:26 +0200
Message-Id: <1333564586-3259-2-git-send-email-javier@dowhile0.org>
X-Mailer: git-send-email 1.7.7.6
In-Reply-To: <1333564586-3259-1-git-send-email-javier@dowhile0.org>
References: <1333564586-3259-1-git-send-email-javier@dowhile0.org>
X-Gm-Message-State: ALoCoQnpqKmpwAEd8KPARCrKLSGTmk0Ya+L63c+vCYWbyTKJkSltXVzBtHyHKTEY+t7xPt0wZ8XW
Cc: linaro-kernel@lists.linaro.org, linaro-dev@lists.linaro.org,
 patches@linaro.org,
 =?UTF-8?q?Enric=20Balletb=C3=B2=20i=20Serra?= <eballetbo@gmail.com>,
 Russell King <linux@arm.linux.org.uk>,
 Javier Martinez Canillas <javier@dowhile0.org>, linux-omap@vger.kernel.org,
 linux-arm-kernel@lists.infradead.org
X-BeenThere: linux-arm-kernel@lists.infradead.org
X-Mailman-Version: 2.1.14
Precedence: list
List-Id: <linux-arm-kernel.lists.infradead.org>
List-Unsubscribe: <http://lists.infradead.org/mailman/options/linux-arm-kernel>, 
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=unsubscribe>
List-Archive: <http://lists.infradead.org/pipermail/linux-arm-kernel/>
List-Post: <mailto:linux-arm-kernel@lists.infradead.org>
List-Help: <mailto:linux-arm-kernel-request@lists.infradead.org?subject=help>
List-Subscribe: <http://lists.infradead.org/mailman/listinfo/linux-arm-kernel>, 
 <mailto:linux-arm-kernel-request@lists.infradead.org?subject=subscribe>
MIME-Version: 1.0
Content-Type: text/plain; charset="us-ascii"
Content-Transfer-Encoding: 7bit
Sender: linux-arm-kernel-bounces@lists.infradead.org
Errors-To: linux-arm-kernel-bounces+thomas.petazzoni=free-electrons.com@lists.infradead.org

IGEP-based boards can have two different flash memories, a OneNAND or
a NAND device. The boot configuration pins (sys_boot) are used to
specify which memory is available.

Also, this patch removes unnecesary code for registering the OneNAND.

Signed-off-by: Javier Martinez Canillas <javier@dowhile0.org>
---
 arch/arm/mach-omap2/board-igep0020.c |   75 ++++++++++++++--------------------
 1 files changed, 31 insertions(+), 44 deletions(-)

diff --git a/arch/arm/mach-omap2/board-igep0020.c b/arch/arm/mach-omap2/board-igep0020.c
index 930c0d3..4af615a 100644
--- a/arch/arm/mach-omap2/board-igep0020.c
+++ b/arch/arm/mach-omap2/board-igep0020.c
@@ -24,6 +24,8 @@
 #include <linux/i2c/twl.h>
 #include <linux/mmc/host.h>
 
+#include <linux/mtd/nand.h>
+
 #include <asm/mach-types.h>
 #include <asm/mach/arch.h>
 
@@ -39,6 +41,8 @@
 #include "hsmmc.h"
 #include "sdram-numonyx-m65kxxxxam.h"
 #include "common-board-devices.h"
+#include "board-flash.h"
+#include "control.h"
 
 #define IGEP2_SMSC911X_CS       5
 #define IGEP2_SMSC911X_GPIO     176
@@ -60,6 +64,10 @@
 #define IGEP3_GPIO_LED1_RED	16
 #define IGEP3_GPIO_USBH_NRESET  183
 
+#define IGEP_SYSBOOT_MASK           0x1f
+#define IGEP_SYSBOOT_NAND           0x0f
+#define IGEP_SYSBOOT_ONENAND        0x10
+
 /*
  * IGEP2 Hardware Revision Table
  *
@@ -110,8 +118,10 @@ static void __init igep2_get_revision(void)
 	gpio_free(IGEP2_GPIO_LED1_RED);
 }
 
-#if defined(CONFIG_MTD_ONENAND_OMAP2) || \
-	defined(CONFIG_MTD_ONENAND_OMAP2_MODULE)
+#if defined(CONFIG_MTD_ONENAND_OMAP2) ||		\
+	defined(CONFIG_MTD_ONENAND_OMAP2_MODULE) ||	\
+	defined(CONFIG_MTD_NAND_OMAP2) ||		\
+	defined(CONFIG_MTD_NAND_OMAP2_MODULE)
 
 #define ONENAND_MAP             0x20000000
 
@@ -123,7 +133,7 @@ static void __init igep2_get_revision(void)
  * So MTD regards it as 4KiB page size and 256KiB block size 64*(2*2048)
  */
 
-static struct mtd_partition igep_onenand_partitions[] = {
+static struct mtd_partition igep_flash_partitions[] = {
 	{
 		.name           = "X-Loader",
 		.offset         = 0,
@@ -151,50 +161,27 @@ static struct mtd_partition igep_onenand_partitions[] = {
 	},
 };
 
-static struct omap_onenand_platform_data igep_onenand_data = {
-	.parts = igep_onenand_partitions,
-	.nr_parts = ARRAY_SIZE(igep_onenand_partitions),
-	.dma_channel	= -1,	/* disable DMA in OMAP OneNAND driver */
-};
-
-static struct platform_device igep_onenand_device = {
-	.name		= "omap2-onenand",
-	.id		= -1,
-	.dev = {
-		.platform_data = &igep_onenand_data,
-	},
-};
+static inline u32 igep_get_sysboot_value(void)
+{
+	return omap_ctrl_readl(OMAP343X_CONTROL_STATUS) & IGEP_SYSBOOT_MASK;
+}
 
 static void __init igep_flash_init(void)
 {
-	u8 cs = 0;
-	u8 onenandcs = GPMC_CS_NUM + 1;
-
-	for (cs = 0; cs < GPMC_CS_NUM; cs++) {
-		u32 ret;
-		ret = gpmc_cs_read_reg(cs, GPMC_CS_CONFIG1);
-
-		/* Check if NAND/oneNAND is configured */
-		if ((ret & 0xC00) == 0x800)
-			/* NAND found */
-			pr_err("IGEP: Unsupported NAND found\n");
-		else {
-			ret = gpmc_cs_read_reg(cs, GPMC_CS_CONFIG7);
-			if ((ret & 0x3F) == (ONENAND_MAP >> 24))
-				/* ONENAND found */
-				onenandcs = cs;
-		}
-	}
-
-	if (onenandcs > GPMC_CS_NUM) {
-		pr_err("IGEP: Unable to find configuration in GPMC\n");
-		return;
-	}
-
-	igep_onenand_data.cs = onenandcs;
-
-	if (platform_device_register(&igep_onenand_device) < 0)
-		pr_err("IGEP: Unable to register OneNAND device\n");
+	u32 mux;
+	mux = igep_get_sysboot_value();
+
+	if (mux == IGEP_SYSBOOT_NAND) {
+		pr_info("IGEP: initializing NAND memory device\n");
+		board_nand_init(igep_flash_partitions,
+				ARRAY_SIZE(igep_flash_partitions),
+				0, NAND_BUSWIDTH_16);
+	} else if (mux == IGEP_SYSBOOT_ONENAND) {
+		pr_info("IGEP: initializing OneNAND memory device\n");
+		board_onenand_init(igep_flash_partitions,
+				   ARRAY_SIZE(igep_flash_partitions), 0);
+	} else
+		pr_err("IGEP: Flash: unsupported sysboot sequence found\n");
 }
 
 #else
-- 
1.7.7.6


_______________________________________________
linux-arm-kernel mailing list
linux-arm-kernel@lists.infradead.org
http://lists.infradead.org/mailman/listinfo/linux-arm-kernel
